### DISCUSSION TITLE
Enforce validation

### DISCUSSION THREAD

**Discussion by TitoSniberb** [react-hook-form/react-hook-form#9319] [Ideas]
If I have an input with a maxLength: 25 rule for example, there should also be an option to not allow new characters into the input until there is less than 25. Right now to be able to do that you have to use the Controller component and that breaks the efficiency of using react hook form. If I have a form with 20 different inputs, I have to have 20 different Controllers with all the boilerplate code just so I can pass it a custom onChange function? Thats not really nice

**Comment by Moshyfawn** [comment]
It's uncontrolled vs controlled inputs trade-off. The former has it's performance benefits as the source of truth for the input's value is the input. The trade-off is you can't "tell" the input it's value _(format / manipulate value)_. If you need to do so, you want the controlled input API _(`Controller` / `useController`)_.

In regards to boilerplate, it depends on how you've setup your components / App. You can create a reusable input with `useController` hook to integrate with your forms easily without writing too much code every time you setup the form.

In regards to uncontrolled inputs and the RHF benefits, the `register` method and the whole mental model of uncontrolled inputs is built around embracing the native Web forms API. For example, to limit your input characters length, there's [`maxlength` attribute](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/maxlength) you can pass to your native text input.
